File: src-tauri/src/care_instructions.rs
// specs.rs
// Ojos Project
//
// Stores care instructions
use crate::structs::CareInstruction;
use chrono::Local;
use rusqlite::{named_params, Connection};
use tauri::{AppHandle, Manager};

pub fn add_care_instruction(
    app: &AppHandle,
    title: String,
    content: String,
    frequency: Option<String>,
    added_by: String,
) -> CareInstruction {
    let app_data_dir = app.path().app_data_dir().unwrap();
    let conn = Connection::open(app_data_dir.join("iris.db")).unwrap();
    let ts = Local::now().timestamp();

    let ci = CareInstruction {
        title,
        content,
        frequency,
        added_by,
        last_updated: ts,
    };

    conn.execute("INSERT INTO care_instruction(title, content, frequency, added_by, last_updated) VALUES (?1, ?2, ?3, ?4, ?5)", (&ci.title, &ci.content, &ci.frequency, &ci.added_by, &ci.last_updated)).unwrap();
    ci
}

pub fn update_care_instructions(
    app: &AppHandle,
    title: String,
    content: String,
    frequency: Option<String>,
    added_by: String,
) -> CareInstruction {
    let app_data_dir = app.path().app_data_dir().unwrap();
    let conn = Connection::open(app_data_dir.join("iris.db")).unwrap();
    let ts = Local::now().timestamp();

    let ci = CareInstruction {
        title,
        content,
        frequency,
        added_by,
        last_updated: ts,
    };

    conn.execute(
        "UPDATE care_instruction SET title=:title, content=:content, frequency=:frequency, added_by=:added_by, last_updated=:last_updated WHERE title=:title",
        named_params! {
            ":title": &ci.title,
            ":content": &ci.content,
            ":frequency": &ci.frequency,
            ":added_by": &ci.added_by,
            ":last_updated": &ci.last_updated
        },
    )
    .unwrap();

    ci
}

pub fn get_all_care_instructions(app: &AppHandle) -> Vec<CareInstruction> {
    let app_data_dir = app.path().app_data_dir().unwrap();
    let conn = Connection::open(app_data_dir.join("iris.db")).unwrap();

    let mut stmt = conn
        .prepare("SELECT * FROM care_instruction ORDER BY last_updated DESC")
        .unwrap();
    let matched_ci = stmt
        .query_map([], |row| {
            Ok(CareInstruction {
                title: row.get(0)?,
                content: row.get(1)?,
                frequency: row.get(2)?,
                added_by: row.get(3)?,
                last_updated: row.get(4)?,
            })
        })
        .unwrap();

    let mut vec_to_return: Vec<CareInstruction> = vec![];
    for ci in matched_ci {
        vec_to_return.push(ci.unwrap());
    }

    vec_to_return
}

File: src/utils/care_instructions.ts
import { CareInstruction } from "@/types/care-instructions";
import { connectUserDB } from "./database";
import { stamp, unix_timestamp } from "./stamp";

export async function getCareInstructions(): Promise<CareInstruction[]> {
  const conn = await connectUserDB();
  return await conn.select(
    "SELECT * FROM care_instruction ORDER BY last_updated DESC",
  );
}

export async function getCareInstruction(
  id: string,
): Promise<CareInstruction | undefined> {
  const conn = await connectUserDB();
  const instructions: CareInstruction[] = await conn.select(
    "SELECT * FROM care_instruction ORDER BY last_updated DESC",
  );

  return instructions.find((instruction) => instruction.id === id);
}

export async function createCareInstruction(
  title: string,
  content: string,
): Promise<CareInstruction> {
  const conn = await connectUserDB();
  const [timestamp, uuid] = await stamp();

  const instruction: CareInstruction = {
    id: uuid,
    title,
    content,
    last_updated: timestamp,
  };

  await conn.execute(
    "INSERT INTO care_instruction(id, title, content, last_updated) VALUES ($1, $2, $3, $4)",
    [
      instruction.id,
      instruction.title,
      instruction.content,
      instruction.last_updated,
    ],
  );

  return instruction;
}

export async function updateCareInstruction(
  id: string,
  title: string,
  content: string,
): Promise<CareInstruction> {
  const conn = await connectUserDB();
  const ts = await unix_timestamp();

  const instruction: CareInstruction = {
    id,
    title,
    content,
    last_updated: ts,
  };

  await conn.execute(
    "UPDATE care_instruction SET title=$2, content=$3, last_updated=$4 WHERE id=$1",
    [
      instruction.id,
      instruction.title,
      instruction.content,
      instruction.last_updated,
    ],
  );

  return instruction;
}

export async function deleteCareInstruction(id: string) {
  const conn = await connectUserDB();

  await conn.execute("DELETE FROM care_instruction WHERE id=$1", [id]);
}

export async function neighboringIds(id: string): Promise<string[]> {
  const instructions = await getCareInstructions();
  const totalLength = instructions.length;

  const index = instructions.findIndex((i) => i.id === id);
  const previous =
    index === 0 ? instructions[totalLength - 1].id : instructions[index - 1].id;
  const next =
    index === totalLength - 1 ? instructions[0].id : instructions[index + 1].id;

  return [previous, next];
}
